src/coreclr/jit/promotiondecomposition.cpp

```diff
@@ -978,23 +861,225 @@ class DecompositionPlan
+        //
+        unsigned FindRegularlyPromotedField(unsigned offs, Compiler* comp)
+        {
+            if (m_local == nullptr)
+            {
+                return BAD_VAR_NUM;
+            }
+
+            LclVarDsc* lclDsc  = comp->lvaGetDesc(m_local);
+            unsigned   lclOffs = m_local->GetLclOffs() + offs;
+
+            if (!lclDsc->lvPromoted)
+            {
+                return BAD_VAR_NUM;
+            }
+
+            return comp->lvaGetFieldLocal(lclDsc, lclOffs);
+        }
+
+        //------------------------------------------------------------------------
+        // GrabAddress:
+        //   Create a derived access of the address at a specified offset. Only
+        //   valid if this represents an indirection. This will decrement the
+        //   number of expected derived accesses that can be created.
+        //
+        // Parameters:
+        //   offs - offset
+        //   comp - Compiler instance
+        //
+        // Returns:
+        //   Node computing the address.
+        //
+        GenTree* GrabAddress(unsigned offs, Compiler* comp)
+        {
+            assert((m_addr != nullptr) && (m_numUsesLeft > 0));
+            m_numUsesLeft--;
+
+            GenTree* addrUse;
+            if (m_numUsesLeft == 0)
+            {
+                // Last use of the address, reuse the node.
+                addrUse = m_addr;
+            }
+            else
+            {
+                addrUse = comp->gtCloneExpr(m_addr);
+            }
+
+            target_ssize_t fullOffs = m_addrBaseOffs + (target_ssize_t)offs;
+            if ((fullOffs != 0) || (m_addrBaseOffsFldSeq != nullptr))
+            {
+                GenTreeIntCon* offsetNode = comp->gtNewIconNode(fullOffs, TYP_I_IMPL);
+                offsetNode->gtFieldSeq    = m_addrBaseOffsFldSeq;
+
+                var_types addrType = varTypeIsGC(addrUse) ? TYP_BYREF : TYP_I_IMPL;
+                addrUse            = comp->gtNewOperNode(GT_ADD, addrType, addrUse, offsetNode);
+            }
+
+            return addrUse;
+        }
+
+#ifdef DEBUG
+        //------------------------------------------------------------------------
+        // CheckFullyUsed:
+        //   If this is an indirection, verify that it was accessed exactly the
+        //   expected number of times.
+        //
+        void CheckFullyUsed()
+        {
+            assert((m_addr == nullptr) || (m_numUsesLeft == 0));
+        }
+#endif
+
+    private:
+        //------------------------------------------------------------------------
+        // GetIndirFlags:
+        //   Get the flags to set on a new indir.
+        //
+        // Parameters:
+        //   type - Type of the indirection
+        //
+        // Returns:
+        //   Flags to set.
+        //
+        GenTreeFlags GetIndirFlags(var_types type)
+        {
+            if (genTypeSize(type) == 1)
+            {
+                return m_indirFlags & ~GTF_IND_UNALIGNED;
```

Comment:
Maybe it's better to be done inside `gtNewIndir` itself?

Reply:
Not sure, creating these derived indirections probably doesn't happen in any other place than decomposition and block morph. And even if it does we probably don't even need to remove this flag... I'm just being overly cautious here.